### PR TITLE
feat: [MR-634] Introduce msg_limit for application subnets in payload builder

### PR DESCRIPTION
**Pull request by stiegerc**
The `slice.header.begin` of a stream slice we induct is used for garbage collecting signals. Therefore the number of signals after inducting a stream slice is `slice.end - slice.header.begin` (note: `slice.header.begin` <= `slice.begin`). Or in other words, if we want to cap the number of signals in the reverse stream at `MAX_STREAM_MESSAGES` (10k), we cannot induct messages beyond `slice.header.begin + MAX_STREAM_MESSAGES`.

This PR changes the logic of `Payload::take_prefix` such that the above mentioned limit is respected.

Note that this can not lead to a lock scenario, because a case where we can't include anymore messages can always be resolved by moving the `begin` in the incoming `slice`, i.e. by garbage collecting on the sending subnet.

### CODE REVIEW

**Comment by stiegerc** [comment]
`ExpectedIndices::message_index` is the `signals_end` of the outgoing stream because it is assigned as such [here](https://sourcegraph.com/github.com/dfinity/ic@0b886f258ddea45421bd7edaa11b698144da528d/-/blob/rs/xnet/payload_builder/src/lib.rs?L451).

**Comment by alin-at-dfinity** [comment]
> After inducting a stream slice, the number of signals in the outgoing stream is equal to **_number_of_signals_after_gcing_ + _number_of_messages_inducted_**. So if we want to cap the number of signals in a stream at 50k, we can induct no more than **_50k_ - _number_of_signals_after_gcing_** number of messages.
>
> Gc'ing signals is done using [begin](https://sourcegraph.com/github.com/dfinity/ic@6ed86361e01cc843eb55d044b904d8b1e8dc6b8f/-/blob/rs/messaging/src/routing/stream_handler.rs?L528) in the stream slice, so **_number_of_signals_after_gcing_ = _outgoing_stream_signals_end_ - _incoming_slice_begin_**; the [signals_end](https://sourcegraph.com/github.com/dfinity/ic@0b886f258ddea45421bd7edaa11b698144da528d/-/blob/rs/xnet/payload_builder/src/lib.rs?L451) of the outgoing stream is given by `message_index` of `ExpectedIndices`.

I find this very hard to parse. A simpler way of putting it is that if we want to limit the outgoing signals to 50k, then we should not induct messages beyond index `incoming_stream_begin + 50k` (do note that that's "incoming stream begin", which is different from "incoming slice begin"). How many (or rather which) messages we should induct directly follows from the above (from wherever we are now to `incoming_stream_begin + 50k`).

File: rs/xnet/payload_builder/src/lib.rs [hunk 1]
```diff
@@ -879,11 +883,24 @@ impl XNetPayloadBuilderImpl {
 ///
 /// In order to prevent mutual stalling, only applies to incoming NNS
 /// streams; and to `Application`-subnet-to-`System`-subnet streams.
-pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usize> {
+pub fn get_msg_limit(
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:886]: The doc comment needs updating.

On a second thought, I would rather have a separate function (or just inline code) to compute an upper bound for `slice.end` (which is just `stream.begin + 50k`, nothing fancy). As is, I'm not 100% sure about this, but we may not even enforce the 50k limit for application subnet streams on system subnets.

It would be a lot easier if we kept this as is; and then, where we call it, simply took the minimum of whatever this returns and `stream_begin + 50k - expected.message_index`. Then you're obviously under both limits at all times.

File: rs/xnet/payload_builder/src/lib.rs [hunk 2]
```diff
@@ -879,11 +883,24 @@ impl XNetPayloadBuilderImpl {
...
+    subnet_id: SubnetId,
+    state: &ReplicatedState,
+    expected: Option<&ExpectedIndices>,
+    slice_begin: StreamIndex,
+) -> Option<usize> {
+    let get_limit = || -> Option<usize> {
+        const MAX_SIGNALS_STREAM_COUNT: usize = 50_000;
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:893]: This should be more than a constant hidden down in some function. If nothing else, it should be a documented constant somewhere at the top of the crate / module.

Also `MAX_SIGNALS_STREAM_COUNT` is hard to read. `MAX_SIGNAL_COUNT`? `MAX_STREAM_SIGNALS`? `MAX_STREAM_SIGNAL_COUNT`? (Although the latter is kind of redundant).

File: rs/xnet/payload_builder/src/lib.rs [hunk 3]
```diff
@@ -879,11 +883,24 @@ impl XNetPayloadBuilderImpl {
...
+        let delta = expected.map_or(0, |expected| {
+            (expected.message_index.get() - slice_begin.get()) as usize
+        });
+        Some(MAX_SIGNALS_STREAM_COUNT - delta)
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:897]: I'd rather put this as in my first comment on the PR: `stream_begin + 50k - expected.message_index()`.

File: rs/xnet/payload_builder/src/lib.rs [hunk 4]
```diff
@@ -879,11 +883,24 @@ impl XNetPayloadBuilderImpl {
 ///
 /// In order to prevent mutual stalling, only applies to incoming NNS
 /// streams; and to `Application`-subnet-to-`System`-subnet streams.
-pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usize> {
+pub fn get_msg_limit(
+    subnet_id: SubnetId,
+    state: &ReplicatedState,
+    expected: Option<&ExpectedIndices>,
+    slice_begin: StreamIndex,
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:890]: Nit: This is `stream_begin`, not `slice_begin`.

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 5]
```diff
@@ -1158,6 +1162,7 @@ impl CertifiedSlicePool {
             Some(entry) => entry,
             None => return Ok(None),
         };
+        let msg_limit = get_msg_limit(subnet_id, state, begin, slice.payload.header.begin());
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/certified_slice_pool.rs:1165]: If you had a separate function from `get_msg_limit()` (or just an inline calculation) you wouldn't need to pass down `state` all the way here. The 50k limit has nothing to do with the state, it is strictly determined by the incoming stream and `begin`.

File: rs/xnet/payload_builder/src/lib.rs (hunk 1 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs]: Made it its own function `get_signal_limit()` that follows the logic you suggest here.

File: rs/xnet/payload_builder/src/lib.rs (hunk 4 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs]: This comment I don't understand. I called it `slice_begin` because that is what is used for [garbage collecting signals](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/messaging/src/routing/stream_handler.rs?L528).

**Comment by stiegerc** [comment]
AFAICT `msg_limit` in [take_prefix()](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/xnet/payload_builder/src/certified_slice_pool.rs?L764) is now `Some(_)` always so the `Option<_>` can be removed, but I left it in since I am not sure whether we may want to keep it as it is.

File: rs/xnet/payload_builder/src/lib.rs [hunk 6]
```diff
@@ -910,6 +926,24 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
...
+/// signals the reverse stream will contain after inducting the stream slice.
+///
+/// This is computed such that the stream index of the last message is at most `slice_begin +
+/// 50_000`. Since `slice_begin` is used for garbage collecting signals, this ensures at most
+/// `50_000` signals in the reverse stream.
+pub fn get_signal_limit(begin: Option<&ExpectedIndices>, slice_begin: StreamIndex) -> usize {
+    // TODO: Replace the 50_000 with a constant. This should be the same constant used in the
+    // stream builder: `MAX_STREAM_MESSAGES`.
+    begin.map_or(50_000, |begin| {
+        let slice_end_index = slice_begin + 50_000.into();
```
Comment (derlerd-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:939]: Nit: I'd call this
```suggestion
        let max_slice_end_index = slice_begin + 50_000.into();
```
so that it is clear that this is the last index we would possibly want to accept.

File: rs/xnet/payload_builder/src/lib.rs [hunk 7]
```diff
@@ -910,6 +926,24 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
...
+/// Calculates an upper bound for how many messages can be included into a block based on how many
+/// signals the reverse stream will contain after inducting the stream slice.
+///
+/// This is computed such that the stream index of the last message is at most `slice_begin +
+/// 50_000`. Since `slice_begin` is used for garbage collecting signals, this ensures at most
+/// `50_000` signals in the reverse stream.
+pub fn get_signal_limit(begin: Option<&ExpectedIndices>, slice_begin: StreamIndex) -> usize {
+    // TODO: Replace the 50_000 with a constant. This should be the same constant used in the
+    // stream builder: `MAX_STREAM_MESSAGES`.
+    begin.map_or(50_000, |begin| {
```
Comment (derlerd-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:938]: I have a hard time following why `begin` is optional here? Shouldn't we always know the begin of the slice we are about to put in a block?

Also looking at the code above it seems that we are passing `Some(...)` at all call sites.

File: rs/xnet/payload_builder/src/lib.rs [hunk 8]
```diff
@@ -712,6 +712,22 @@ impl XNetPayloadBuilderImpl {
                     ));
                 }
             }
+
+            // Ensure the signal limit is respected.
+            let signal_limit = get_signal_limit(Some(expected), slice.header().begin());
+            if messages.len() > signal_limit {
```
Comment (derlerd-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:718]: The logic why this can be done separately from the check above about `msg_limit` is that we will include the `min` of `signal_limit` and `msg_limit` so for an honestly built block we know that both must pass right?

File: rs/xnet/payload_builder/src/lib.rs (hunk 4 above)
Comment (alin-at-dfinity) [reply rs/xnet/payload_builder/src/lib.rs]: As you observed on Slack, `stream_slice.header().begin()` is the beginning of the stream, not of the slice. The header in question is the stream header. A slice consists of the stream header and some range of messages. The stream header would be the same regardless of which messages are included in the slice (or if no messages are included).

I.e. this is the stream begin, as certified by the remote subnet. Not the slice begin.

File: rs/xnet/payload_builder/tests/certified_slice_pool.rs [hunk 9]
```diff
@@ -131,11 +131,16 @@ proptest! {
...
+            if let Some(prefix) = prefix.as_ref() {
+                if let Some(msg_limit) = msg_limit {
+                    assert!(testing::slice_len(prefix) <= msg_limit);
+                }
+                if let Some(byte_limit) = byte_limit {
+                    assert!(prefix.count_bytes() <= byte_limit);
+                }
+                if let Some(slice_end) = testing::slice_end(prefix) {
+                    assert!(slice_end <= testing::stream_begin(prefix) + StreamIndex::new(MAX_STREAM_MESSAGES as u64));
+                }
```
Comment (stiegerc) [inline rs/xnet/payload_builder/tests/certified_slice_pool.rs:143]: This is more for the sake of completion since it is never called with number anywhere near large enough to meaningfully check this. Since it is a proptest, I'd rather not extend this to include 10k messages in every single iteration and I'd rather add a small unit test to check for the signals limit.

File: rs/xnet/payload_builder/src/lib.rs [hunk 10]
```diff
@@ -910,6 +932,29 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
...
+///
+/// - `slice_begin` is the index of the first message in the stream slice.
+/// - `stream_begin` is the `begin` in the `StreamHeader` contained in the (same) stream slice.
+///    It reflects the status on the remote subnet as far as we know at present. Up to this index
+///    signals can be gc'ed in the reverse stream.
+///
+/// XNet traffic between honest subnets implies `slice_begin >= stream_begin` since the remote
+/// subnet can only gc messages once this subnet reports corresponding signals (generated by
+/// inducting messages from a stream slice).
+pub fn get_signal_limit(stream_begin: StreamIndex, slice_begin: Option<StreamIndex>) -> usize {
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs]: That's exactly what I just came up with as well <:O)

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 11]
```diff
@@ -1159,6 +1159,17 @@ impl CertifiedSlicePool {
             None => return Ok(None),
         };
 
+        let signal_limit = get_signal_limit(
+            slice.payload.header.begin(),
+            slice
+                .payload
+                .messages
+                .as_ref()
+                .map(|messages| messages.begin()),
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/xnet/payload_builder/src/certified_slice_pool.rs]: Actually I just moved this whole computation further down into [take_prefix on Payload](https://sourcegraph.com/github.com/dfinity/ic@c0b2ec6580bdaf80047b9e7b125a86eab424adff/-/blob/rs/xnet/payload_builder/src/certified_slice_pool.rs?L380). Since it's now defined on the slice itself (`Payload` is part of `UnpackedStreamSlice`) it should be ok to use just `slice.begin` and `slice.header.begin` right? I mean if it's an invalid slice this would seem to be the wrong place to reject it anyway.

File: rs/xnet/payload_builder/src/lib.rs (hunk 7 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs]: I moved the new limit all the way down [here](https://sourcegraph.com/github.com/dfinity/ic@c0b2ec6580bdaf80047b9e7b125a86eab424adff/-/blob/rs/xnet/payload_builder/src/certified_slice_pool.rs?L380), where the `Option` is unwrapped anyway.

One other advantage should be that I can test this is in isolation without an entire pool of > 10k messages in it. I have to give that a try first though.

File: rs/xnet/payload_builder/src/lib.rs [hunk 12]
```diff
@@ -231,6 +231,9 @@ impl XNetPayloadBuilderMetrics {
     }
 }
 
+// TODO: Unify this constant and `MAX_STREAM_MESSAGES` in the stream builder.
+pub const MAX_STREAM_MESSAGES: usize = 10_000;
```
Comment (stiegerc) [inline rs/xnet/payload_builder/src/lib.rs:235]: Probably not for this MR, but we should either make the constant in `StreamBuilder` pub or else define the constant somewhere else or perhaps even better make this an argument passed in from the outside.

This way we could make it part of some config and thus configurable in state machine tests. It would be great if we could put these limits to something much smaller for testing. The stream size and also this change here won't trigger a crash or anything of the sort of the constant is changed as far as I know.

Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs:235]: Actually even for this change it would be useful to have it configurable in proptests so that we don't have to generate a stream with > 10k messages in it every iteration.

File: rs/xnet/payload_builder/tests/certified_slice_pool.rs [hunk 13]
```diff
@@ -137,6 +137,8 @@ proptest! {
             if let (Some(byte_limit), Some(prefix)) = (byte_limit, prefix.as_ref()) {
                 assert!(prefix.count_bytes() <= byte_limit);
             }
+            // Testing the signal limit is pointless here because it requires very large streams
+            // that would make this test needlessly slow. There is a dedicated test for it.
```
Comment (stiegerc) [inline rs/xnet/payload_builder/tests/certified_slice_pool.rs:141]: See the comment about making the limits configurable. I could easily add it here but I don't want to generate huge streams all the time.

File: rs/xnet/payload_builder/tests/certified_slice_pool.rs [hunk 14]
```diff
@@ -1114,3 +1116,49 @@ proptest! {
...
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(10))]
+
+    // Testing the 'signals limit' (the limit on messages in a slice such that the number of
+    // signals after inducting it is capped) requires streams with thousands of messages in it.
+    //
+    // It is therefore using a reduced number of cases to keep the load within reasonable bounds.
+    #[test]
+    fn pool_take_slice_respects_signal_limit(
```
Comment (stiegerc) [inline rs/xnet/payload_builder/tests/certified_slice_pool.rs:1128]: Made this test before I moved the change with `max_messages_end` down into `Payload::take_prefix()` so it's now perhaps not in the ideal place. But this tests what it should and if we make the `MAX_STREAM_MESSAGES` into an argument instead, I can probably integrate the test for the new limit as a 3 liner elsewhere.

I'd rather keep this as it is for now and think about the other change on a different PR. There is at least one issue with the detection of dishonest subnets due to too many signals. We can (and probably should) change that check to use 10k rather than 200k after this is has been rolled out for a while. But that assumes its a constant...

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 15]
```diff
@@ -377,7 +377,19 @@ impl Payload {
...
+        let max_message_limit = {
+            let messages_begin =
+                self.messages_begin().unwrap_or(self.header.begin()).get() as usize;
+            let max_messages_end = max_messages_end(self.header.begin()).get() as usize;
+            max_messages_end.saturating_sub(messages_begin)
+        };
+        let message_limit = message_limit.map_or(max_message_limit, |message_limit| {
+            message_limit.min(max_message_limit)
+        });
+        //let message_limit = message_limit.unwrap_or(usize::MAX);
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/certified_slice_pool.rs:392]:
```suggestion
```

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 16]
```diff
@@ -377,7 +377,19 @@ impl Payload {
...
         byte_limit: Option<usize>,
     ) -> CertifiedSliceResult<(Option<Self>, Option<Self>)> {
-        let message_limit = message_limit.unwrap_or(usize::MAX);
+        // `messages_end` may not exceed a certain `max_messages_end` in order to cap the number of
+        // signals in the reverse stream. Calculate the maximum number of messages we can include
+        // in the slice such that this is respected.
+        let max_message_limit = {
+            let messages_begin =
+                self.messages_begin().unwrap_or(self.header.begin()).get() as usize;
+            let max_messages_end = max_messages_end(self.header.begin()).get() as usize;
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/certified_slice_pool.rs:386]: Personal opinion: I'd consider naming the function and variable something like `max_message_index`. "Max messages end" requires an extra level of indirection: "messages end" is the end of the slice, we don't want the slice to go beyond some index. But basically, all we want to say is "don't induct anything past this message index". (That this results in a limit on a slice's `messages_end` is a side-effect.)

File: rs/xnet/payload_builder/src/lib.rs [hunk 17]
```diff
@@ -526,6 +529,7 @@ impl XNetPayloadBuilderImpl {
         }
 
         if !reject_signals.is_empty() {
+            // TODO: MR-635 Change this check to use `MAX_STREAM_MESSAGES` instead.
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:532]:
```suggestion
            // TODO(MR-635): Change this check to use `MAX_STREAM_MESSAGES` instead.
```

File: rs/xnet/payload_builder/src/lib.rs [hunk 18]
```diff
@@ -903,6 +923,16 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
     }
 }
 
+/// Calculates an upper bound for up to which stream index messages can be included into a block
+/// based on how many signals the reverse stream will contain after inducting the stream slice.
```
Comment (alin-at-dfinity) [inline rs/xnet/payload_builder/src/lib.rs:927]:
```suggestion
/// The stream index up to which messages can be inducted while limiting the
/// number of signals in the reverse stream to `MAX_STREAM_MESSAGES`.
```

File: rs/xnet/payload_builder/src/lib.rs (hunk 12 above)
Comment (alin-at-dfinity) [reply rs/xnet/payload_builder/src/lib.rs:228]: Agreed about making this configurable, either only on paper (by having a constant somewhere in `/rs/replica` and passing it in the constructors of all relevant components); or by actually adding a field somewhere under `rs/config` (and passing it the same way).

In either case (but especially in the latter) we must be careful with how we validate that this limit is correctly enforced.

**Review by alin-at-dfinity — APPROVED** [review]

File: rs/xnet/payload_builder/src/certified_slice_pool.rs (hunk 16 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/certified_slice_pool.rs]: `max_messages_index()` it is.

File: rs/xnet/payload_builder/src/lib.rs (hunk 12 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs:228]: Made a [ticket](https://dfinity.atlassian.net/jira/software/c/projects/MR/boards/223?selectedIssue=MR-636).

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 19]
```diff
@@ -377,7 +377,18 @@ impl Payload {
         message_limit: Option<usize>,
         byte_limit: Option<usize>,
     ) -> CertifiedSliceResult<(Option<Self>, Option<Self>)> {
-        let message_limit = message_limit.unwrap_or(usize::MAX);
+        // `messages_end` may not exceed a certain `max_messages_index` in order to cap the number
```
Comment (oggy-dfin) [inline rs/xnet/payload_builder/src/certified_slice_pool.rs]: Could we change the comment above to something like:
> Takes a slice prefix whose estimated size meets the explicit limits below, as well as an implicit limit on the number of messages which ensures that the number of signals in the reverse stream stays bounded.

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 20]
```diff
@@ -377,7 +377,18 @@ impl Payload {
...
+        // in the slice such that this is respected.
+        let max_message_limit = {
+            let messages_begin =
+                self.messages_begin().unwrap_or(self.header.begin()).get() as usize;
+            let max_messages_index = max_messages_index(self.header.begin()).get() as usize;
+            max_messages_index.saturating_sub(messages_begin)
+        };
+        let message_limit = message_limit.map_or(max_message_limit, |message_limit| {
+            message_limit.min(max_message_limit)
+        });
```
Comment (oggy-dfin) [inline rs/xnet/payload_builder/src/certified_slice_pool.rs]: This seems a bit more convoluted than necessary, how about something like:
```suggestion
        let max_message_limit = {
            let nr_existing_signals = self.messages_begin.map_or(0, |mb| mb - self.header.begin()) as usize;
            MAX_STREAM_MESSAGES - nr_existing_signals
        };
        let message_limit = message_limit.unwrap_or(max_message_limit).min(max_message_limit);
```
(probably doesn't compile, but you get the idea)

File: rs/xnet/payload_builder/src/certified_slice_pool.rs (hunk 16 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/certified_slice_pool.rs]: Done.

File: rs/xnet/payload_builder/src/lib.rs [hunk 21]
```diff
@@ -526,6 +530,7 @@ impl XNetPayloadBuilderImpl {
         }
 
         if !reject_signals.is_empty() {
+            // TODO(MR-635): Change this check to use `MAX_STREAM_MESSAGES` instead.
```
Comment (oggy-dfin) [inline rs/xnet/payload_builder/src/lib.rs:533]: I find this TODO confusing since the check does seem to use `MAX_STREAM_MESSAGES` below, but I guess you know what it's about, if you're gonna sort this out soon it's fine by me.

File: rs/xnet/payload_builder/src/certified_slice_pool.rs [hunk 22]
```diff
@@ -377,7 +377,18 @@ impl Payload {
...
+        // in the slice such that this is respected.
+        let max_message_limit = {
+            let messages_begin =
+                self.messages_begin().unwrap_or(self.header.begin()).get() as usize;
+            let max_message_index = max_message_index(self.header.begin()).get() as usize;
+            max_message_index.saturating_sub(messages_begin)
+        };
+        let message_limit = message_limit.map_or(max_message_limit, |message_limit| {
+            message_limit.min(max_message_limit)
+        });
```
Comment (oggy-dfin) [inline rs/xnet/payload_builder/src/certified_slice_pool.rs:391]: This seems more complicated than necessary, how about something like:
```suggestion
        let max_message_limit = {
            let nr_existing_signals =  self.messages_begin().map_or(0, |mr| mr - self.header.begin()) as usize;
            MAX_STREAM_MESSAGES - nr_existing_signals 
        };
        let message_limit = message_limit.unwrap_or(max_message_limit).min(max_message_limit);
```
(probably doesn't compile, but you get the gist; the last line is arguably not much simpler than what you had before, feel free to ignore that part)

File: rs/xnet/payload_builder/src/impl_tests.rs [hunk 23]
```diff
@@ -510,6 +510,96 @@ async fn validate_slice_above_msg_limit() {
...
+        let stream = generate_stream(&StreamConfig {
+            message_begin: STREAM_BEGIN,
+            message_end: MAX_STREAM_END + 10,
+            signal_end: SIGNALS_END,
+        });
+
+        // Helper for validating a generated slice from `SUBNET_1` taken from `stream`
+        // with messages starting end ending at the given indices.
+        let validate_slice = |slice_begin: u64, slice_end: u64, state| {
+            let slice = stream.slice(slice_begin.into(), Some((slice_end - slice_begin) as usize));
```
Comment (oggy-dfin) [inline rs/xnet/payload_builder/src/impl_tests.rs:554]: If the stream is oversized (but not the slice), do we generate a warning anywhere?

**Review by oggy-dfin — COMMENTED** [review]
LGTM overall, I left a comment about the index juggling that I think can be simplified. What's unclear to me is whether we issue a warning if the stream itself is above `MAX_STREAM_MESSAGES`, IIUC this should not happen with honest subnets?

File: rs/xnet/payload_builder/src/lib.rs (hunk 21 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/lib.rs]: It is indeed. Wrote it differently. What I meant to say is the old check uses 200k signals for the check arguing about how many messages could fit in a stream. But since this PR caps them at 10k, that's what we should use for this check.

**Comment by stiegerc** [comment]
> LGTM overall, I left a comment about the index juggling that I think can be simplified. What's unclear to me is whether we issue a warning if the stream itself is above `MAX_STREAM_MESSAGES`, IIUC this should not happen with honest subnets?

Oversized streams are a normal part of operation, but they can only do that with reject responses. So before reject signals for requests were introduced: If you can't deliver a request for whatever reason, the subnet generated a reject response and pushed it into the stream, this had to be done even if the stream was at capacity, hence the stream got oversized.

Now with reject signals for requests we have a similar situation when we get a reject signal for which we generate a reject response locally. If the canister has migrated in the mean time, we have to push it into the stream to its new subnet instead, even if the stream is full.

File: rs/xnet/payload_builder/src/certified_slice_pool.rs (hunk 19 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/certified_slice_pool.rs]: Forgot to mention this in the comment, good catch.

File: rs/xnet/payload_builder/src/certified_slice_pool.rs (hunk 20 above)
Comment (stiegerc) [reply rs/xnet/payload_builder/src/certified_slice_pool.rs]: I like my version better for a couple of reasons:
- It has only one subtraction in it to worry about.
- It doesn't use the constant `MAX_STREAM_MESSAGES` allowing me to encapsulate this in the function `max_message_index()`.
- It follows a logic of 3 places on the stream index axis I find easier to understand.
I had the same mental picture that your suggestion is based on initially, but the picture of a stream index axis seems cleaner to me since it also fits better with the check in `validate_slice()`.

I mean something like this.
```
--------|------------------|-------------------------|---------------------> stream index
 stream_begin        messages_begin             max_message_index
```
So we have a `max_message_index` up to which we can have signals. After gc'ing up to `stream_begin`, we have signals from there to `messages_begin`. Therefore we can produce `max_message_index - messages_begin` more signals. The `saturating_sub()` is in there for the case where you decrease the number of signals in streams in which case we could have `max_message_index < messages_begin`.

Actually why don't I just put this into the comment above the `max_message_index()` function?

File: rs/xnet/payload_builder/src/lib.rs [hunk 24]
```diff
@@ -903,6 +925,16 @@ pub fn get_msg_limit(subnet_id: SubnetId, state: &ReplicatedState) -> Option<usi
...
 
+/// The stream index up to which messages can be inducted while limiting the
+/// number of signals in the reverse stream to `MAX_STREAM_MESSAGES`.
+///
+/// `stream_begin` is the `begin` in the `StreamHeader` contained in the (same) stream slice.
+///  It reflects the status on the remote subnet as far as we know at present. Up to this index
+///  signals can be gc'ed in the reverse stream.
+pub fn max_message_index(stream_begin: StreamIndex) -> StreamIndex {
+    stream_begin + (MAX_STREAM_MESSAGES as u64).into()
+}
```
Comment (stiegerc) [inline rs/xnet/payload_builder/src/lib.rs:936]: I was just wondering what would happen if we got a malicious `stream_begin` here, i.e. what if this wraps around?
First of all it would mean that `max_message_index` << `messages_begin` in which case we just wouldn't include any messages in slices.
But there are also sanity checks for this in `validate_slice()`, for [begin and end in the header](https://sourcegraph.com/github.com/dfinity/ic@5c85e7d4570d9636f896ec6fdb0e507ff4e01d08/-/blob/rs/xnet/payload_builder/src/lib.rs?L619) and also to ensure `messages_begin` is [within this interval](https://sourcegraph.com/github.com/dfinity/ic@5c85e7d4570d9636f896ec6fdb0e507ff4e01d08/-/blob/rs/xnet/payload_builder/src/lib.rs?L619).

So unless the stream indices are very close to `u64::MAX` in which case all kinds of other stuff would also break, this is not an issue. We'll never get there though; it would take thousands of years.

File: rs/xnet/payload_builder/src/certified_slice_pool.rs (hunk 20 above)
Comment (oggy-dfin) [reply rs/xnet/payload_builder/src/certified_slice_pool.rs]: Personally I didn't find the `max_message_index` concept particularly helpful, when looking at the code I had to unfold the `max_message_index`definition to figure out what's going on and how it's linked to the motivation of "we don't want too many signals", both here and in `validate_slice`. So I'd find it easier to just get rid of it completely. But you'll have to stare at this code more than I will so I won't push for it, feel free to resolve.

**Comment by oggy-dfin** [comment]
Thanks! I missed the case of a migrating canister, a warning doesn't make sense if this can happen as part of normal operation (even if subnet splitting is a bit of special case)

**Review by oggy-dfin — APPROVED** [review]

File: rs/xnet/payload_builder/src/lib.rs (hunk 24 above)
Comment (alin-at-dfinity) [reply rs/xnet/payload_builder/src/lib.rs:936]: This is one of the reasons why I was arguing for using `ExpectedInfices` for this, i.e. the value that we know to be correct, as opposed to a value fed to us by a potentially malicious subnet.

Even though the latter is probably validated again and again, including in the block maker, so by the time it gets to us it's just as safe to use. But (to some extent) it removes the need to wonder about "what if this index was invalid".
